## Flight Price Prediction (EDA And Feature Engineering)

In [1]:
## Data Set link: https://github.com/krishnaik06/5-Days-Live-EDA-and-Feature-Engineering/tree/main/Flight%20Prediction

In [165]:
# importing basics libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [223]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

In [166]:
# Data Loading
train_df = pd.read_excel(r"C:\Users\asinga\OneDrive - PVH Corp\Documents\iNeuron Data Science\Flight Price Prediction\Data_Train.xlsx")

In [167]:
train_df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [168]:
test_df = pd.read_excel(r"C:\Users\asinga\OneDrive - PVH Corp\Documents\iNeuron Data Science\Flight Price Prediction\Test_set.xlsx")

In [169]:
test_df.head()

Airline Date_of_Journey    Source Destination            Route  \
0        Jet Airways       6/06/2019     Delhi      Cochin  DEL → BOM → COK   
1             IndiGo      12/05/2019   Kolkata    Banglore  CCU → MAA → BLR   
2        Jet Airways      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
3  Multiple carriers      21/05/2019     Delhi      Cochin  DEL → BOM → COK   
4           Air Asia      24/06/2019  Banglore       Delhi        BLR → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops              Additional_Info  
0    17:30  04:25 07 Jun  10h 55m      1 stop                      No info  
1    06:20         10:20       4h      1 stop                      No info  
2    19:15  19:00 22 May  23h 45m      1 stop  In-flight meal not included  
3    08:00         21:00      13h      1 stop                      No info  
4    23:55  02:45 25 Jun   2h 50m    non-stop                      No info

In [170]:
final_df = pd.concat([train_df,test_df],ignore_index=True)

In [171]:
final_df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info    Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897.0  
1    05:50         13:15   7h 25m     2 stops         No info   7662.0  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882.0  
3    18:05         23:30   5h 25m      1 stop         No info   6218.0  
4    16:50         21:35   4h 45m      1 stop         No info  13302.0

In [172]:
final_df.tail()

Airline Date_of_Journey   Source Destination  \
13349          Air India       6/06/2019  Kolkata    Banglore   
13350             IndiGo      27/03/2019  Kolkata    Banglore   
13351        Jet Airways       6/03/2019    Delhi      Cochin   
13352          Air India       6/03/2019    Delhi      Cochin   
13353  Multiple carriers      15/06/2019    Delhi      Cochin   

                 Route Dep_Time  Arrival_Time Duration Total_Stops  \
13349  CCU → DEL → BLR    20:30  20:25 07 Jun  23h 55m      1 stop   
13350        CCU → BLR    14:20         16:55   2h 35m    non-stop   
13351  DEL → BOM → COK    21:50  04:25 07 Mar   6h 35m      1 stop   
13352  DEL → BOM → COK    04:00         19:15  15h 15m      1 stop   
13353  DEL → BOM → COK    04:55         19:15  14h 20m      1 stop   

      Additional_Info  Price  
13349         No info    NaN  
13350         No info    NaN  
13351         No info    NaN  
13352         No info    NaN  
13353         No info    NaN

In [173]:
final_df.shape

(13354, 11)

In [174]:
final_df.columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [175]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13354 entries, 0 to 13353
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
dtypes: float64(1), object(10)
memory usage: 1.1+ MB


In [176]:
final_df['Date_of_Journey'].str.split('/').str[0]

0        24
1         1
2         9
3        12
4        01
         ..
13349     6
13350    27
13351     6
13352     6
13353    15
Name: Date_of_Journey, Length: 13354, dtype: object

In [177]:
## Feature Engineering Process
final_df['Day'] = final_df['Date_of_Journey'].str.split('/').str[0]
final_df['Month'] = final_df['Date_of_Journey'].str.split('/').str[1]
final_df['Year'] = final_df['Date_of_Journey'].str.split('/').str[2]


In [178]:
final_df.head(2)

Airline Date_of_Journey    Source Destination                  Route  \
0     IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1  Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info   Price Day  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info  3897.0  24   
1    05:50         13:15   7h 25m     2 stops         No info  7662.0   1   

  Month  Year  
0    03  2019  
1    05  2019

In [179]:
#final_df['Day'] = final_df['Date_of_Journey'].apply(lambda x:x.split('/')[0])

In [180]:
final_df['Day'] = final_df['Day'].astype(int)
final_df['Month'] = final_df['Month'].astype(int)
final_df['Year'] = final_df['Year'].astype(int)

In [181]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13354 entries, 0 to 13353
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Date_of_Journey  13354 non-null  object 
 2   Source           13354 non-null  object 
 3   Destination      13354 non-null  object 
 4   Route            13353 non-null  object 
 5   Dep_Time         13354 non-null  object 
 6   Arrival_Time     13354 non-null  object 
 7   Duration         13354 non-null  object 
 8   Total_Stops      13353 non-null  object 
 9   Additional_Info  13354 non-null  object 
 10  Price            10683 non-null  float64
 11  Day              13354 non-null  int32  
 12  Month            13354 non-null  int32  
 13  Year             13354 non-null  int32  
dtypes: float64(1), int32(3), object(10)
memory usage: 1.3+ MB


In [182]:
final_df.drop('Date_of_Journey', axis = 1, inplace=True)

In [183]:
final_df.head(1)

Airline    Source Destination      Route Dep_Time  Arrival_Time Duration  \
0  IndiGo  Banglore   New Delhi  BLR → DEL    22:20  01:10 22 Mar   2h 50m   

  Total_Stops Additional_Info   Price  Day  Month  Year  
0    non-stop         No info  3897.0   24      3  2019

In [184]:
final_df['Arrival_Time'] = final_df['Arrival_Time'].str.split(' ').str[0]

In [185]:
##final_df[Arrival_Time].apply(lambda x:x.split(' ').str[0])

In [186]:
final_df.head(1)

Airline    Source Destination      Route Dep_Time Arrival_Time Duration  \
0  IndiGo  Banglore   New Delhi  BLR → DEL    22:20        01:10   2h 50m   

  Total_Stops Additional_Info   Price  Day  Month  Year  
0    non-stop         No info  3897.0   24      3  2019

In [187]:
## Cheking Null values
final_df.isnull().sum()

Airline               0
Source                0
Destination           0
Route                 1
Dep_Time              0
Arrival_Time          0
Duration              0
Total_Stops           1
Additional_Info       0
Price              2671
Day                   0
Month                 0
Year                  0
dtype: int64

In [188]:
final_df['Arrival_hour'] = final_df['Arrival_Time'].str.split(':').str[0]
final_df['Arrival_min'] = final_df['Arrival_Time'].str.split(':').str[1]

In [189]:
final_df.head(2)

Airline    Source Destination                  Route Dep_Time  \
0     IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1  Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   

  Arrival_Time Duration Total_Stops Additional_Info   Price  Day  Month  Year  \
0        01:10   2h 50m    non-stop         No info  3897.0   24      3  2019   
1        13:15   7h 25m     2 stops         No info  7662.0    1      5  2019   

  Arrival_hour Arrival_min  
0           01          10  
1           13          15

In [190]:
final_df['Arrival_hour']  =final_df['Arrival_hour'].astype(int)
final_df['Arrival_min']  =final_df['Arrival_min'].astype(int)

In [191]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13354 entries, 0 to 13353
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Dep_Time         13354 non-null  object 
 5   Arrival_Time     13354 non-null  object 
 6   Duration         13354 non-null  object 
 7   Total_Stops      13353 non-null  object 
 8   Additional_Info  13354 non-null  object 
 9   Price            10683 non-null  float64
 10  Day              13354 non-null  int32  
 11  Month            13354 non-null  int32  
 12  Year             13354 non-null  int32  
 13  Arrival_hour     13354 non-null  int32  
 14  Arrival_min      13354 non-null  int32  
dtypes: float64(1), int32(5), object(9)
memory usage: 1.3+ MB


In [192]:
final_df.drop('Arrival_Time', axis=1, inplace=True)

In [193]:
final_df.head(1)

Airline    Source Destination      Route Dep_Time Duration Total_Stops  \
0  IndiGo  Banglore   New Delhi  BLR → DEL    22:20   2h 50m    non-stop   

  Additional_Info   Price  Day  Month  Year  Arrival_hour  Arrival_min  
0         No info  3897.0   24      3  2019             1           10

In [194]:
final_df['Dep_hour'] = final_df['Dep_Time'].str.split(':').str[0]
final_df['Dep_min'] = final_df['Dep_Time'].str.split(':').str[1]

In [195]:
final_df['Dep_hour']  =final_df['Dep_hour'].astype(int)
final_df['Dep_min']  =final_df['Dep_min'].astype(int)

In [196]:
final_df.head(1)

Airline    Source Destination      Route Dep_Time Duration Total_Stops  \
0  IndiGo  Banglore   New Delhi  BLR → DEL    22:20   2h 50m    non-stop   

  Additional_Info   Price  Day  Month  Year  Arrival_hour  Arrival_min  \
0         No info  3897.0   24      3  2019             1           10   

   Dep_hour  Dep_min  
0        22       20

In [197]:
final_df.drop('Dep_Time', axis=1, inplace=True)

In [198]:
final_df.head(1)

Airline    Source Destination      Route Duration Total_Stops  \
0  IndiGo  Banglore   New Delhi  BLR → DEL   2h 50m    non-stop   

  Additional_Info   Price  Day  Month  Year  Arrival_hour  Arrival_min  \
0         No info  3897.0   24      3  2019             1           10   

   Dep_hour  Dep_min  
0        22       20

In [199]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13354 entries, 0 to 13353
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13354 non-null  object 
 1   Source           13354 non-null  object 
 2   Destination      13354 non-null  object 
 3   Route            13353 non-null  object 
 4   Duration         13354 non-null  object 
 5   Total_Stops      13353 non-null  object 
 6   Additional_Info  13354 non-null  object 
 7   Price            10683 non-null  float64
 8   Day              13354 non-null  int32  
 9   Month            13354 non-null  int32  
 10  Year             13354 non-null  int32  
 11  Arrival_hour     13354 non-null  int32  
 12  Arrival_min      13354 non-null  int32  
 13  Dep_hour         13354 non-null  int32  
 14  Dep_min          13354 non-null  int32  
dtypes: float64(1), int32(7), object(7)
memory usage: 1.2+ MB


In [200]:
final_df['Total_Stops'].unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [201]:
final_df['Total_Stops'] = final_df['Total_Stops'].map({'non-stop':0,'2 stops':1,'1 stop':2,'3 stops':3,'4 stops':4,'nan':1})

In [202]:
final_df[final_df['Total_Stops'].isnull()]

Airline Source Destination Route Duration  Total_Stops  \
9039  Air India  Delhi      Cochin   NaN  23h 40m          NaN   

     Additional_Info   Price  Day  Month  Year  Arrival_hour  Arrival_min  \
9039         No info  7480.0    6      5  2019             9           25   

      Dep_hour  Dep_min  
9039         9       45

In [203]:
final_df.drop('Route', axis=1, inplace=True)

In [204]:
final_df.head()

Airline    Source Destination Duration  Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m          0.0         No info   
1    Air India   Kolkata    Banglore   7h 25m          1.0         No info   
2  Jet Airways     Delhi      Cochin      19h          1.0         No info   
3       IndiGo   Kolkata    Banglore   5h 25m          2.0         No info   
4       IndiGo  Banglore   New Delhi   4h 45m          2.0         No info   

     Price  Day  Month  Year  Arrival_hour  Arrival_min  Dep_hour  Dep_min  
0   3897.0   24      3  2019             1           10        22       20  
1   7662.0    1      5  2019            13           15         5       50  
2  13882.0    9      6  2019             4           25         9       25  
3   6218.0   12      5  2019            23           30        18        5  
4  13302.0    1      3  2019            21           35        16       50

In [205]:
final_df['Additional_Info'].unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [206]:
final_df['duration_hour'] = final_df['Duration'].str.split(' ').str[0].str.split('h').str[0]

In [207]:
final_df.drop(6474, axis=0, inplace=True)
final_df.drop(13343, axis=0, inplace=True)

In [208]:
final_df[final_df['duration_hour']=='5m']

Empty DataFrame
Columns: [Airline, Source, Destination, Duration, Total_Stops, Additional_Info, Price, Day, Month, Year, Arrival_hour, Arrival_min, Dep_hour, Dep_min, duration_hour]
Index: []

In [209]:
final_df['duration_hour'] = final_df['duration_hour'].astype(int)

In [210]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13352 entries, 0 to 13353
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Airline          13352 non-null  object 
 1   Source           13352 non-null  object 
 2   Destination      13352 non-null  object 
 3   Duration         13352 non-null  object 
 4   Total_Stops      13351 non-null  float64
 5   Additional_Info  13352 non-null  object 
 6   Price            10682 non-null  float64
 7   Day              13352 non-null  int32  
 8   Month            13352 non-null  int32  
 9   Year             13352 non-null  int32  
 10  Arrival_hour     13352 non-null  int32  
 11  Arrival_min      13352 non-null  int32  
 12  Dep_hour         13352 non-null  int32  
 13  Dep_min          13352 non-null  int32  
 14  duration_hour    13352 non-null  int32  
dtypes: float64(2), int32(8), object(5)
memory usage: 1.2+ MB


In [211]:
final_df.drop('Duration', axis=1, inplace=True)

In [212]:
final_df.head(1)

Airline    Source Destination  Total_Stops Additional_Info   Price  Day  \
0  IndiGo  Banglore   New Delhi          0.0         No info  3897.0   24   

   Month  Year  Arrival_hour  Arrival_min  Dep_hour  Dep_min  duration_hour  
0      3  2019             1           10        22       20              2

In [213]:
final_df['Airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [214]:
## Label Encoding
from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()

In [215]:
final_df['Airline'] = labelencoder.fit_transform(final_df['Airline'])
final_df['Source'] = labelencoder.fit_transform(final_df['Source'])
final_df['Destination'] = labelencoder.fit_transform(final_df['Destination'])
final_df['Additional_Info'] = labelencoder.fit_transform(final_df['Additional_Info'])

In [216]:
final_df.shape

(13352, 14)

In [217]:
final_df.head(2)

Airline  Source  Destination  Total_Stops  Additional_Info   Price  Day  \
0        3       0            5          0.0                8  3897.0   24   
1        1       3            0          1.0                8  7662.0    1   

   Month  Year  Arrival_hour  Arrival_min  Dep_hour  Dep_min  duration_hour  
0      3  2019             1           10        22       20              2  
1      5  2019            13           15         5       50              7

In [218]:
ohe.fit_transform(np.array(final_df['Airline']).reshape(-1,1))

<13352x12 sparse matrix of type '<class 'numpy.float64'>'
	with 13352 stored elements in Compressed Sparse Row format>

In [220]:
final_df = pd.get_dummies(final_df, columns=['Airline', 'Source', 'Destination', 'Additional_Info'], drop_first=True).fillna(0).astype(int)

In [221]:
final_df.head()

Total_Stops  Price  Day  Month  Year  Arrival_hour  Arrival_min  Dep_hour  \
0            0   3897   24      3  2019             1           10        22   
1            1   7662    1      5  2019            13           15         5   
2            1  13882    9      6  2019             4           25         9   
3            2   6218   12      5  2019            23           30        18   
4            2  13302    1      3  2019            21           35        16   

   Dep_min  duration_hour  ...  Destination_5  Additional_Info_1  \
0       20              2  ...              1                  0   
1       50              7  ...              0                  0   
2       25             19  ...              0                  0   
3        5              5  ...              0                  0   
4       50              4  ...              1                  0   

   Additional_Info_2  Additional_Info_3  Additional_Info_4  Additional_Info_5  \
0                  0                  0                  0                  0   
1                  0                  0                  0                  0   
2                  0                  0                  0                  0   
3                  0                  0                  0                  0   
4                  0                  0                  0                  0   

   Additional_Info_6  Additional_Info_7  Additional_Info_8  Additional_Info_9  
0                  0                  0                  1                  0  
1                  0                  0                  1                  0  
2                  0                  0                  1                  0  
3                  0                  0                  1                  0  
4                  0                  0                  1                  0  

[5 rows x 39 columns]

In [222]:
final_df.columns

Index(['Total_Stops', 'Price', 'Day', 'Month', 'Year', 'Arrival_hour',
       'Arrival_min', 'Dep_hour', 'Dep_min', 'duration_hour', 'Airline_1',
       'Airline_2', 'Airline_3', 'Airline_4', 'Airline_5', 'Airline_6',
       'Airline_7', 'Airline_8', 'Airline_9', 'Airline_10', 'Airline_11',
       'Source_1', 'Source_2', 'Source_3', 'Source_4', 'Destination_1',
       'Destination_2', 'Destination_3', 'Destination_4', 'Destination_5',
       'Additional_Info_1', 'Additional_Info_2', 'Additional_Info_3',
       'Additional_Info_4', 'Additional_Info_5', 'Additional_Info_6',
       'Additional_Info_7', 'Additional_Info_8', 'Additional_Info_9'],
      dtype='object')

## Model Bulding

In [224]:
X = final_df.drop(columns=["Price"])
y = final_df["Price"]

In [225]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [228]:
## 1. Linear Regression
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

y_pred_lr = linear_model.predict(X_test)
mse_lr = mean_squared_error(y_test, y_pred_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print(f"Linear Regression - MSE: {mse_lr}, R²: {r2_lr}")

Linear Regression - MSE: 21226772.820725303, R²: 0.32842994822591054


In [229]:
## 2. Random Forest Regressor
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)
mse_rf = mean_squared_error(y_test, y_pred_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print(f"Random Forest - MSE: {mse_rf}, R²: {r2_rf}")

Random Forest - MSE: 24120125.620847143, R²: 0.2368904049241919


In [230]:
## 3. Gradient Boosting Regressor
from sklearn.ensemble import GradientBoostingRegressor

gb_model = GradientBoostingRegressor(random_state=42)
gb_model.fit(X_train, y_train)

y_pred_gb = gb_model.predict(X_test)
mse_gb = mean_squared_error(y_test, y_pred_gb)
r2_gb = r2_score(y_test, y_pred_gb)

print(f"Gradient Boosting - MSE: {mse_gb}, R²: {r2_gb}")

Gradient Boosting - MSE: 20466733.290852036, R²: 0.35247598625242504


In [231]:
## 4. XGBoost Regressor
from xgboost import XGBRegressor

xgb_model = XGBRegressor(random_state=42)
xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print(f"XGBoost - MSE: {mse_xgb}, R²: {r2_xgb}")

XGBoost - MSE: 21503469.77021397, R²: 0.3196758438567149


In [232]:
## 5. LightGBM Regressor
from lightgbm import LGBMRegressor

lgbm_model = LGBMRegressor(random_state=42)
lgbm_model.fit(X_train, y_train)

y_pred_lgbm = lgbm_model.predict(X_test)
mse_lgbm = mean_squared_error(y_test, y_pred_lgbm)
r2_lgbm = r2_score(y_test, y_pred_lgbm)

print(f"LightGBM - MSE: {mse_lgbm}, R²: {r2_lgbm}")


[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.001818 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 172
[LightGBM] [Info] Number of data points in the train set: 10681, number of used features: 27
[LightGBM] [Info] Start training from score 7242.549293
LightGBM - MSE: 21193899.753420785, R²: 0.329469982323327


In [233]:
## Compare Model Performance

performance = {
    "Model": ["Linear Regression", "Random Forest", "Gradient Boosting", "XGBoost", "LightGBM"],
    "MSE": [mse_lr, mse_rf, mse_gb, mse_xgb, mse_lgbm],
    "R²": [r2_lr, r2_rf, r2_gb, r2_xgb, r2_lgbm],
}

performance_df = pd.DataFrame(performance)
print(performance_df)


               Model           MSE        R²
0  Linear Regression  2.122677e+07  0.328430
1      Random Forest  2.412013e+07  0.236890
2  Gradient Boosting  2.046673e+07  0.352476
3            XGBoost  2.150347e+07  0.319676
4           LightGBM  2.119390e+07  0.329470


## Step 5: Choose the Best Model
Select the model with the lowest MSE and highest R².

## Gradient Boosting:

MSE (Mean Square Error): 20,466,730 (best).
R²: 0.352476 (best).
Gradient Boosting is the best-performing model here, capturing more variance in the target variable.